processor/schemaprocessor/internal/migrate/attributes.go

```diff
@@ -0,0 +1,108 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migrate // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/migrate"
+
+import (
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/alias"
+)
+
+// AttributeKey is used to allow strings and type defined
+// strings to be used within the creation of creating an `AttributeChangeSet`
+type AttributeKey interface {
+	~string
+}
+
+// AttributeChangeSet represents an unscoped entry that can be applied.
+//
+// The listed changes are duplicated twice
+// to allow for simplified means of transition to or from a revision.
+type AttributeChangeSet struct {
+	updates  map[alias.AttributeKey]alias.AttributeKey
+	rollback map[alias.AttributeKey]alias.AttributeKey
+}
+
+// OrderedAttributeChangeSets allows for `AttributeChangeSet`
+// to be chained together as they are defined within the schema
+// and be applied sequentially to ensure deterministic behavior.
+type AttributeChangeSetSlice []*AttributeChangeSet
```

Comment:
Nit: this probably can be:
```suggestion
type AttributeChangeSetSlice []AttributeChangeSet
```

Reply:
I'd like to keep it as pointers to avoid copying entire values onto the stack for each processing step.

I am conscious that optimising early might be a bad idea without actual information of user feedback.